sycl/doc/extensions/Uniform/Uniform.asciidoc

```diff
@@ -0,0 +1,193 @@
+the value of a variable is the same for all work-items in a sub-group.
+Asserting uniformity when it does not hold can lead to undefined behavior.
+
+An implementation is not required to perform any analysis to prove or disprove
+the uniformity of a variable passed to the `sycl::ext::oneapi::experimental::uniform`
+constructor, but it is highly recommended that an implementation issue a
+diagnostic if it is known that an expression can never be uniform.
+
+Several classes in SYCL encapsulate information that is known to vary across
+work-items; instances of such classes can never be uniform. An implementation
+must issue a diagnostic if the `sycl::ext::oneapi::experimental::uniform` class template is
+instantiated with any of the following types: `sycl::item`, `sycl::nd_item`,
+`sycl::h_item`, `sycl::group`, `sycl::sub_group` and `sycl::nd_range`.
+
+[source, c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+template <class T>
+class uniform {
+  explicit uniform(T x) noexcept;
+  operator T() const;
+};
+
+} // namespace experimental
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+----
+
+The `sycl::ext::oneapi::experimental::uniform` wrapper class does not support arithmetic
+operators or provide access to the underlying data, and is intended to be
+used to represent that the value of a variable is uniform at a specific point
+(i.e. when the wrapper class is constructed). The variable can be implicitly
+converted back to an instance of `T`, but doing so is not guaranteed to
+retain any uniformity information. The implicit conversion is free to return
+a value taken from any work-item in the sub-group; if the input to the
+constructor was not uniform, or the constructor was not encountered in
+converged control flow, then the resulting value is undefined.
+
+== Example Usage
+
+This non-normative section shows some example usages of the extension.
+
+Overloading a function for uniform and non-uniform arguments:
+```c++
+// If ptr is assumed non-uniform, use atomics to update it
+template <typename T>
+void update(sub_group sg, T* ptr, T x) {
+  sycl::atomic_ref<T, sycl::memory_order::relaxed, sycl::memory_scope::device>(ptr) += x;
+}
+
+// If ptr is asserted to be uniform, use a sub-group reduce first
+template <typename T>
+void update(sub_group sg, sycl::ext::oneapi::experimental::uniform<T*> ptr, T x) {
+  T sum = sycl::reduce_over_group(sg, x, std::plus<>());
+  if (sg.leader()) {
+    sycl::atomic_ref<T, sycl::memory_order::relaxed, sycl::memory_scope::device>(ptr) += sum;
+  }
+}
+```
+
+Asserting that all work-items in the sub-group take the same branch:
+```c++
+bool condition = ...;
+if (sycl::ext::oneapi::experimental::uniform(condition)) {
+  ...
+}
+```
+
+Asserting that all work-items in the sub-group access the same memory location:
+```c++
+float x = array[sycl::ext::oneapi::experimental::uniform(index)];
+```
```

Comment:
Assuming that the uniformity should be processed by the compiler back end, what is your suggestion about IR produced by all the statements above where uniform is used..

Reply:
That's an implementation detail, and outside of the specification.

That said... remember that SYCL `uniform` does not imply that the variable must be stored differently, and so there is no need for a compiler to work backwards from a `uniform` declaration to examine all prior uses of a variable.  In fact, there is no need for a compiler to do anything at all based on a `uniform` declaration, and this is the expected implementation for some architectures (e.g. NVIDIA GPUs).

If a SYCL implementation wants to use a `uniform` declaration to decorate the variable with some special metadata that propagates to future uses of the variable, it is free to do so.  If it wants to work backwards from the declaration and prove (retroactively) that certain usages of the variable must also have been uniform, it is free to do that too.  But that is not required.